src/stitching/addTypenameForFragments.ts

```diff
@@ -0,0 +1,52 @@
+import { ASTNode, GraphQLSchema, visit, FieldNode, Kind } from 'graphql';
+
+export default function addTypenameForFragments<T extends ASTNode>(
+  node: T,
+  schema: GraphQLSchema,
+): T {
+  return visit(node, {
+    [Kind.FIELD]: {
+      leave(field: FieldNode) {
+        if (field.selectionSet) {
+          const selections = field.selectionSet.selections;
+
+          let hasFragment = false;
+          let hasTypename = false;
+
+          selections.forEach(selection => {
+            if (
+              selection.kind === Kind.INLINE_FRAGMENT ||
+              selection.kind === Kind.FRAGMENT_SPREAD
+            ) {
+              hasFragment = true;
+            } else if (selection.name.value === '__typename') {
+              hasTypename = true;
+            }
+          });
+
+          if (hasFragment && !hasTypename) {
+            return {
+              ...field,
+              selectionSet: {
+                ...field.selectionSet,
+                selections: [
+                  ...field.selectionSet.selections,
+                  {
```

Comment:
I still think it's easier to just not go through the whole selections, but rather just inserts the fragment that will have __typename.